File: abc338/c/main.py
def max_meals(N, Q, A, B):
    dp = [0] * (sum(Q) + 1)  # 各材料の合計量に基づいた配列を用意

    for i in range(N):
        for j in range(sum(Q), -1, -1):
            if j >= A[i]:
                dp[j] = max(dp[j], dp[j - A[i]] + 1)  # 料理Aを作る場合
            if j >= B[i]:
                dp[j] = max(dp[j], dp[j - B[i]] + 1)  # 料理Bを作る場合

    return max(dp)

# 入力を受け取る
N = int(input())
Q = list(map(int, input().split()))
A = list(map(int, input().split()))
B = list(map(int, input().split()))

# 関数を呼び出して結果を表示
print(max_meals(N, Q, A, B))

File: abc338/c/main.rb
def max_meals(n, q, a, b)
  # 各材料で作れる料理Aと料理Bの最大人数を計算
  max_meals_a = Array.new(n) { |i| a[i] != 0 ? q[i] / a[i] : 0 }
  max_meals_b = Array.new(n) { |i| b[i] != 0 ? q[i] / b[i] : 0 }

  # 全ての組み合わせを試して、最大人数を見つける
  max_total_meals = 0
  (0..max_meals_a.min).each do |meals_a|
    (0..max_meals_b.min).each do |meals_b|
      if (0...n).all? { |i| q[i] >= a[i] * meals_a + b[i] * meals_b }
        max_total_meals = [max_total_meals, meals_a + meals_b].max
      end
    end
  end

  max_total_meals
end

# 入力例
n = gets.to_i
q = gets.split.map(&:to_i)
a = gets.split.map(&:to_i)
b = gets.split.map(&:to_i)

# 出力
puts max_meals(n, q, a, b)
